lib/instrumentation/pino/pino.js

```diff
@@ -0,0 +1,152 @@
+   * Pino passes in level as number but our spec needs it to be the label
+   */
+  const levelMap = Object.entries(levelUtils.levels).reduce((levels, level) => {
+    const [label, number] = level
+    levels[number] = label
+    return levels
+  }, {})
+
+  const symbols = shim.require('./lib/symbols')
+
+  /**
+   * Wraps the level cache so we can properly set a formatter to return the
+   * label as level in log line instead of number
+   */
+  shim.wrap(levelUtils, 'genLsCache', function genLsCache(shim, genLevelsCache) {
+    return function wrappedGenLsCache() {
+      const args = shim.argsToArray.apply(shim, arguments)
+      args[0][symbols.formattersSym].level = function nrLevelMapping(label) {
+        return { level: label }
+      }
+      return genLevelsCache.apply(this, args)
+    }
+  })
+
+  shim.wrap(tools, 'asJson', function wrapJson(shim, asJson) {
+    /**
+     * Wraps function in pino that is used to construct/serialize a log
+     * line to json
+     *
+     * @param {object} obj data from mixins
+     * @param {string} msg message of log line
+     * @param {number} num log level as number
+     * @param {string} time formatted snippet of json with time(`,"time":<unix time>`)
+     * @returns {string} serialized log line
+     */
+    return function wrappedAsJson() {
+      // overriding the symbol that defines the key for message(pino defaults this to `msg`)
+      this[symbols.messageKeySym] = 'message'
+
+      const args = shim.argsToArray.apply(shim, arguments)
+
+      /**
+       * changing the label from time to timestamp
+       * and assign unix timestamp to match our specification
+       */
+      args[3] = `,"timestamp":${Date.now()}`
+
+      if (isMetricsEnabled(config)) {
+        const level = args[2]
+        incrementLoggingLinesMetrics(levelMap[level], metrics)
+      }
+
+      if (isLogForwardingEnabled(config, agent)) {
+        const metadata = agent.getLinkingMetadata()
+        reformatLogLine(args[0], metadata)
+      } else if (isLocalDecoratingEnabled(config)) {
+        args[1] += agent.getNRLinkingMetadata()
+      }
+
+      /**
+       * must call original asJson to allow pino
+       * to construct the entire log line before we
+       * add to the log aggregator
+       */
+      const logLine = asJson.apply(this, args)
+
+      if (isLogForwardingEnabled(config, agent)) {
+        agent.logs.add(logLine)
+      }
+
+      return logLine
+    }
+  })
+}
+
+/**
+ * reformats error and assigns NR context data
+ * to log line
+ *
+ * @param {object} logLine log line
+ * @param {object} metadata NR context data
+ */
+function reformatLogLine(logLine, metadata) {
+  if (logLine.err) {
+    reformatError(logLine)
+  }
+
+  /**
+   * pino adds this already for us
+   * since asJson manually constructs the json string,
+   * it will have hostname twice if we do not delete ours.
+   */
+  delete metadata.hostname
```

Comment:
Doesn't this depend on the pino configuration?

https://github.com/pinojs/pino/blob/master/docs/api.md#base-object

We should check if it's there beforehand, no?

Reply:
Good point.  I got caught up in trying to override this symbol to force hostname that I forgot to check this